### PR TITLE
add support for `time.Time`

### PR DESCRIPTION
**Pull request by michaellzc**
related https://github.com/hexops/valast/issues/11

TODO

- [x] support `*time.Time`

I am not sure what is the best way to render `*.time.Time`:

✅   generic (~this is not supported given the current target Go version is 1.15 in `go.mod`~)

```go
// an exported func from autogold
func Ptr[V any](v V) *V {
	return &v
}

// render as
valast.Ptr(time.Date(...))
```

❌  inline anonymous fn
```go
// render as
func(t time.Time) *time.Time { return &t }(time.Date(...))
```

thoughts?

- [x] By selecting this checkbox, I agree to license my contributions to this project under the license(s) described in the LICENSE file, and I have the right to do so or have received permission to do so by an employer or client I am producing work for whom has this right.

### Test plan

all tests passed

```
go test                                                                                                                                                                                                                                                                                                       13:00:24
PASS
ok      github.com/hexops/valast        0.960s
```

Tested this change with workspace in https://github.com/sourcegraph/controller, and it seems to work just fine. It was able to automatically wrap `*time.Time` to `Ptr(time.Date())` and valast pkg was imported properly in `_test.go` file as well.

### CODE REVIEW

**Comment by slimsag** [comment]
Thanks for sending @michaellzc - I'd be happy to merge this.

I can offer a few options:

1. I can merge this as-is, without support for `*time.Time` - I'm content with doing that.
2. You can implement `*time.Time` using the generic `Ptr` approach, I would be happy to have that change and require the latest-and-greatest Go version in go.mod.

I'll let you decide which one you want to go for depending on how much time you want to spend here.

**Comment by michaellzc** [comment]
> 2\. You can implement `*time.Time` using the generic `Ptr` approach, I would be happy to have that change and require the latest-and-greatest Go version in go.mod.

😛 I need `*time.Time` for my own use case, so I'll continue on option 2 with a twist:

```
// an exported func from autogold
func TimePtr(t time.Time) *time.Time {
	return &t
}

// render as
autogold.TimePtr(time.Date(...))
```

`SomethingPtr` is quite a common pattern in many Go projects, e.g., `k8s.io/utils/pointer`, `github.com/aws/jsii-runtime-go`.

it's not elegant, but we can avoid breaking backward compatibility unless we have to.

**Comment by slimsag** [comment]
@michaellzc I'm not a fan of that `TimePtr` change - that feels awkward and a bit cluttery of the API space. I have zero qualms about breaking compatibility, so please do the `Ptr` approach unless there's something that makes that much more difficult

File: valast.go [hunk 1]
```diff
@@ -606,6 +612,15 @@ func computeAST(v reflect.Value, opt *Options, cycleDetector *cycleDetector, pro
 		}
 		return basicLit(vv, token.STRING, "string", strconv.Quote(v.String()), opt.withUnqualify(), typeExprCache)
 	case reflect.Struct:
+		// special handling for common structs from stdlib
+		// that only contiains unexported fields
```
Comment (slimsag) [inline valast.go:616]:
```suggestion
		// that only contain unexported fields
```

**Comment by slimsag** [comment]
@michaellzc not sure if you'd be up for it (fine if not) but maybe you could remove `valast.Addr` in favor of `valast.Ptr`? It's doing the same thing but results in uglier ASTs as it doesn't use generics:

https://github.com/hexops/valast/blob/5de3f44da1923f80e8c0526a257bf8278730f9db/valast.go#L151-L167

No worries if not, just let me know what you decide & I'll merge either way.

File: valast.go [hunk 2]
```diff
@@ -503,15 +505,12 @@ func computeAST(v reflect.Value, opt *Options, cycleDetector *cycleDetector, pro
...
-							Sel: ast.NewIdent("Addr"),
-						},
-						Args: []ast.Expr{elem.AST},
+				AST: &ast.CallExpr{
+					Fun: &ast.SelectorExpr{
+						X:   ast.NewIdent("valast"),
+						Sel: ast.NewIdent("Ptr"),
 					},
-					Type: ptrType.AST,
+					Args: []ast.Expr{elem.AST},
```
Comment (michaellzc) [inline valast.go:513]: 👀

File: valast.go [hunk 3]
```diff
@@ -549,20 +548,23 @@ func computeAST(v reflect.Value, opt *Options, cycleDetector *cycleDetector, pro
...
```
Comment (michaellzc) [inline valast.go:556]: 👀

**Comment by michaellzc** [comment]
> @michaellzc not sure if you'd be up for it (fine if not) but maybe you could remove `valast.Addr` in favor of `valast.Ptr`? It's doing the same thing but results in uglier ASTs as it doesn't use generics:
> 
> https://github.com/hexops/valast/blob/5de3f44da1923f80e8c0526a257bf8278730f9db/valast.go#L151-L167
> 
> No worries if not, just let me know what you decide & I'll merge either way.

refactor from `Addr` to `Ptr` is done.

Trying to understand what is going in `AddrInterface` see if we can get rid of it as well

**Comment by michaellzc** [comment]
> > @michaellzc not sure if you'd be up for it (fine if not) but maybe you could remove `valast.Addr` in favor of `valast.Ptr`? It's doing the same thing but results in uglier ASTs as it doesn't use generics:
> > https://github.com/hexops/valast/blob/5de3f44da1923f80e8c0526a257bf8278730f9db/valast.go#L151-L167
> > 
> > No worries if not, just let me know what you decide & I'll merge either way.
> 
> refactor from `Addr` to `Ptr` is done.
> 
> Trying to understand what is going in `AddrInterface` see if we can get rid of it as well

I may find some existing problems with `AddrInterface`

1. Many tests are not setting options to `ExportedOnly`, and the rendered code contains a private struct `foo` from `internal/test` pkg.

1. Rendered output of `TestPointers/interface` is not valid Go

```go
// input
v := test.Bazer(test.NewBaz())
&v

// output
valast.AddrInterface(test.Bazer{&test.Baz{
	Bam:  (1.34 + 0i),
	zeta: &test.foo{bar: "hello"},
}},
	(*test.Bazer)(nil)).(*test.Bazer)
```

I don't think you can do `test.Bazer{&test.Baz{}}` with an interface in Go.

**Review by slimsag — APPROVED** [review]

**Pull request merged**